dag-cbor-derive/src/gen.rs

```diff
@@ -377,15 +350,21 @@ fn gen_try_read_cbor_union(u: &Union) -> TokenStream {
         }
         UnionRepr::Kinded => {
             let variants = u.variants.iter().map(|s| {
-                let parse = gen_try_read_cbor_struct(s);
+                let parse = gen_decode_struct(s);
                 quote! {
-                    let res: Result<Option<Self>> = #parse;
-                    res?;
+                    let result: Result<Self> = (|| {
+                        #parse
+                    })();
+                    match result {
+                        Ok(res) => return Ok(res),
+                        Err(err) if err.downcast_ref::<UnexpectedCode>().is_some() => {}
+                        Err(err) => return Err(err),
+                    };
```

Comment:
Is this acceptable? we will have allocations when a parse tree fails while parsing kinded unions.

Reply:
For me, yes. That is pretty unlikely to happen on the happy path, unless you have nested kinded unions in the exact right way, right?